examples/cosmos/lightclient/Blockchain.qnt

```diff
@@ -0,0 +1,168 @@
+    // A set of faulty nodes, which can act as validators.
+    // We assume that the set of faulty processes is non-decreasing.
+    // If a process has recovered, it should connect using a different id.
+    var Faulty: Set[NODE]
+
+
+    // The set of all correct nodes in a state
+    def Corr = AllNodes.exclude(Faulty)
+
+
+    // (****************************** BLOCKCHAIN ************************************)
+
+    // the header is still within the trusting period
+    def InTrustingPeriod(header: BlockHeader): bool = refClock < header.time + TRUSTING_PERIOD
+    
+    //  Given a function pVotingPower \in D -> Powers for some D \subseteq AllNodes
+    //  and pNodes \subseteq D, test whether the set pNodes \subseteq AllNodes has
+    //  more than 2/3 of voting power among the nodes in D.
+    pure def TwoThirds(pVS: Set[NODE], pNodes: Set[NODE]): bool = {
+        val TP = size(pVS)
+        val SP = size(pVS.intersect(pNodes))
+        // when thinking in real numbers, not integers: SP > 2.0 / 3.0 * TP 
+        3 * SP > 2 * TP 
+    }
+
+    //  Given a set of FaultyNodes, test whether the voting power of the correct nodes in D
+    //  is more than 2/3 of the voting power of the faulty nodes in D.
+    //  Parameters:
+    //    - pFaultyNodes is a set of nodes that are considered faulty
+    //    - pVS is a set of all validators, maybe including Faulty, intersecting with it, etc.
+    //    - pMaxFaultRatio is a pair <<a, b>> that limits the ratio a / b of the faulty
+    //      validators from above (exclusive)
+    pure def FaultyValidatorsFewerThan(pFaultyNodes: Set[NODE], pVS: Set[NODE], maxRatio: (int, int)): bool = {
+        val FN = pFaultyNodes.intersect(pVS)    // faulty nodes in pNodes
+        val CN = pVS.exclude(pFaultyNodes)      // correct nodes in pNodes
+        val CP = size(CN)                       // power of the correct nodes
+        val FP = size(FN)                       // power of the faulty nodes
+        // CP + FP = TP is the total voting power
+        val TP = CP + FP
+        FP * maxRatio._2 < TP * maxRatio._1
+    }
+
+    //   Can a block be produced by a correct peer, or an authenticated Byzantine peer?
+    def IsLightBlockAllowedByDigitalSignatures(ht: int, block: LightBlock): bool = or {
+        block.header == blockchain.get(ht), // signed by correct and faulty (maybe)
+        and {
+            block.Commits.subseteq(Faulty),
+            block.header.height == ht,
+            block.header.time >= 0 // signed only by faulty
+        }
+    }
+
+    //  Initialize the blockchain to the ultimate height right in the initial states.
+    //  We pick the faulty validators statically, but that should not affect the light client.
+    //  Parameters:
+    //     - pMaxFaultyRatioExclusive is a pair <<a, b>> that bound the number of
+    //         faulty validators in each block by the ratio a / b (exclusive)
+    action InitToHeight(pMaxFaultyRatioExclusive: (int, int)): bool = {
```

Comment:
Just a note here, it is very unlikely that this initialisation would work in the simulator. It is tuned to Apalache.